engine/common/follower/engine.go

```diff
@@ -368,10 +391,24 @@ func (e *Engine) processBlockAndDescendants(ctx context.Context, proposal *messa
 	log.Info().Msg("forwarding block proposal to hotstuff")
 
 	// submit the model to follower for processing
-	e.follower.SubmitProposal(header, parent.View)
+	if inRangeBlockResponse {
+		select {
+		case <-e.follower.SubmitProposal(header, parent.View):
+			// we wait if the block proposal was part of a range block response since
+			// processing block n+1 before block n has finished will fail
```

Comment:
```suggestion
			// after submitting proposal to hotstuff, then hotstuff will start processing block n, and follower
			// engine is concurrently processing block n + 1. 
			// however follower engine will fail to process block n + 1 if block n is not saved in protocol state.
			// Block n is only saved in protocol state when hotstuff finishes processing block n.
			// In order to ensure follower engine don't process block n + 1 too early, we wait until hotstuff finish
			// processing block n.
			// this wait is only needed when processing range block response, since blocks are processed in order.
```